File: ScriptsToProcess/NEW BATCH TO PROCESS/DASH.py
"""
JB Grasshopper Gem - OOP Masterplan JSON Dashboard (V5 - Patched)
================================================================================
Reads a summarized JSON payload from the OOP Masterplan Engine and renders 
a responsive HUD. 
- FIXED: Background box scales correctly with multiple building titles.
- FIXED: Targets are now case-insensitive and display in both Global and Building modes.

INPUTS:
    run, JSON_Payload, GroupByBldg, title, suffix, TargetGlobal, TargetJSON
    size, anchor, font, transparency, ox, oy, fit_padding
================================================================================
"""

import json
import Rhino
import System.Drawing
import scriptcontext as sc

KEY = "JB_GEM_OOP_DASHBOARD_V5"

def format_num(n):
    try: return "{:,.2f}".format(float(n))
    except: return str(n)

# ==============================================================================
# VIEWPORT RENDER CALLBACK
# ==============================================================================
def draw_callback(sender, e):
    data = sc.sticky.get(KEY + "_DATA")
    if not data: return

    display_data = data['display_data'] 
    l_size, l_anchor = data['size'], data['anchor']
    font_face = data['font_face']
    alpha, ox, oy, padding = data['alpha'], data['off_x'], data['off_y'], data['padding']
    
    vp_w, vp_h = e.Viewport.Bounds.Width, e.Viewport.Bounds.Height
    line_h = l_size * 1.6
    title_gap = 20 
    
    max_label_chars = max([len(d['label']) for d in display_data]) if display_data else 10
    value_start_offset = max_label_chars * (l_size * 0.6)
    
    full_lines = [d['label'] + d['val'] for d in display_data]
    max_total_chars = max([len(s) for s in full_lines]) if full_lines else 10
    box_w = (max_total_chars * l_size * 0.55) + (padding * 3.0)
    
    # FIXED: Count EVERY title separator to calculate the exact box height
    title_count = sum(1 for d in display_data if "---" in d['label'] or "===" in d['label'])
    box_h = (len(display_data) * line_h) + (title_count * title_gap) + (padding * 2.5)

    if l_anchor == 1: x, y = vp_w - box_w - ox, oy 
    elif l_anchor == 2: x, y = ox, vp_h - box_h - oy 
    elif l_anchor == 3: x, y = vp_w - box_w - ox, vp_h - box_h - oy 
    else: x, y = ox, oy + 35 

    bg = System.Drawing.Color.FromArgb(int(alpha), 25, 25, 25)
    white = System.Drawing.Color.White
    warning_red = System.Drawing.Color.OrangeRed
    
    rect = System.Drawing.Rectangle(int(x), int(y), int(box_w), int(box_h))
    e.Display.Draw2dRectangle(rect, bg, 1, bg)
    
    current_y = y + padding
    for item in display_data:
        # Skip drawing text for empty spacers, just add height
        if not item['label'].strip() and not item['val'].strip():
            current_y += line_h
            continue
            
        e.Display.Draw2dText(item['label'], white, Rhino.Geometry.Point2d(x + padding, current_y), False, int(l_size), font_face)
        if "---" not in item['label'] and "===" not in item['label']:
            val_p = Rhino.Geometry.Point2d(x + padding + value_start_offset, current_y)
            color = warning_red if item['is_red'] else white
            e.Display.Draw2dText(item['val'], color, val_p, False, int(l_size), font_face)
        
        current_y += (line_h + title_gap) if ("---" in item['label'] or "===" in item['label']) else line_h

# ==============================================================================
# LIFECYCLE MANAGEMENT
# ==============================================================================
def shutdown():
    if KEY in sc.sticky:
        try: Rhino.Display.DisplayPipeline.DrawForeground -= sc.sticky[KEY]
        except: pass
        del sc.sticky[KEY]
    if KEY + "_DATA" in sc.sticky: del sc.sticky[KEY + "_DATA"]
    Rhino.RhinoDoc.ActiveDoc.Views.Redraw()

shutdown()

# ==============================================================================
# DATA PARSING & EXECUTION
# ==============================================================================
run_btn = globals().get('run', False)
json_in = globals().get('JSON_Payload', None)
group_by_bldg = globals().get('GroupByBldg', False)

if run_btn and json_in:
    try:
        data = json.loads(json_in)
        prog_areas = data.get("programs", {})
        total_area = data.get("total_area", 0.0)
        bldg_data = data.get("buildings", {})
        
        # Parse Targets safely (Case-Insensitive matching)
        clean_target_data = {}
        target_json = globals().get('TargetJSON', None)
        if target_json:
            try: 
                raw_targets = json.loads(target_json)
                clean_target_data = {str(k).strip().upper(): float(v) for k, v in raw_targets.items()}
            except: pass
            
        u = str(globals().get('suffix', ""))
        opac = int(max(0.0, min(1.0, float(globals().get('transparency', 0.8) or 0.8))) * 255)
        
        display_data = []
        hud_title = globals().get('title', "MASTERPLAN SUMMARY")
        if hud_title:
            display_data.append({'label': "=== " + str(hud_title).upper() + " ===", 'val': "", 'is_red': False})

        # --- MODE 1: PER BUILDING BREAKDOWN ---
        if group_by_bldg:
            for b_name, b_stats in bldg_data.items():
                display_data.append({'label': "--- " + str(b_name).upper() + " ---", 'val': "", 'is_red': False})
                for p_name, val_area in b_stats["programs"].items():
                    line_label = "  {}: ".format(str(p_name))
                    line_val = "{}{}".format(format_num(val_area), u)
                    display_data.append({'label': line_label, 'val': line_val, 'is_red': False})
                
                b_tot_val = "{}{}".format(format_num(b_stats["total_area"]), u)
                display_data.append({'label': "  SUBTOTAL: ", 'val': b_tot_val, 'is_red': False})
                display_data.append({'label': " ", 'val': " ", 'is_red': False})
            
            # Show Global Targets at the bottom of Building Mode if they exist
            if clean_target_data:
                display_data.append({'label': "--- TARGET TRACKING ---", 'val': "", 'is_red': False})
                for prog_name, val_area in prog_areas.items():
                    target = clean_target_data.get(str(prog_name).strip().upper(), 0.0)
                    if target > 0:
                        pct = (val_area / target) * 100
                        line_label = "  {}: ".format(str(prog_name))
                        line_val = "{}{}".format(format_num(val_area), u)
                        line_val += " / {}{} ({:.1f}%)".format(format_num(target), u, pct)
                        display_data.append({'label': line_label, 'val': line_val, 'is_red': val_area > target})
                display_data.append({'label': " ", 'val': " ", 'is_red': False})
                
            display_data.append({'label': "=" * 15, 'val': "", 'is_red': False})
            
        # --- MODE 2: GLOBAL PROGRAMS ONLY ---
        else:
            for prog_name, val_area in prog_areas.items():
                target = clean_target_data.get(str(prog_name).strip().upper(), 0.0)
                line_label = "{}: ".format(str(prog_name).upper())
                line_val = "{}{}".format(format_num(val_area), u)
                over_target = False
                
                if target > 0:
                    pct = (val_area / target) * 100
                    line_val += " / {}{} ({:.1f}%)".format(format_num(target), u, pct)
                    if val_area > target: over_target = True
                    
                display_data.append({'label': line_label, 'val': line_val, 'is_red': over_target})
            display_data.append({'label': "-" * 15, 'val': "", 'is_red': False})

        # --- GLOBAL MASTERPLAN TOTAL (Always shown at bottom) ---
        tot_is_red = False
        tot_str = "{}{}".format(format_num(total_area), u)
        
        t_glob = globals().get('TargetGlobal', 0)
        if t_glob and t_glob > 0:
            tot_pct = (total_area / t_glob) * 100
            tot_str += " [Target: {}{} ({:.1f}%)]".format(format_num(t_glob), u, tot_pct)
            if total_area > t_glob: tot_is_red = True

        display_data.append({'label': "TOTAL AREA: ", 'val': tot_str, 'is_red': tot_is_red})

        sc.sticky[KEY + "_DATA"] = {
            "display_data": display_data,
            "size": globals().get('size') or 12,
            "anchor": globals().get('anchor') or 0,
            "font_face": globals().get('font') or "Arial",
            "alpha": opac,
            "off_x": float(globals().get('ox') or 20),
            "off_y": float(globals().get('oy') or 20),
            "padding": float(globals().get('fit_padding') or 10)
        }
        
        sc.sticky[KEY] = draw_callback
        Rhino.Display.DisplayPipeline.DrawForeground += draw_callback
        
        mode_msg = "Per-Building Mode" if group_by_bldg else "Global Mode"
        ghenv.Component.Message = "HUD ACTIVE\n{}".format(mode_msg)
        Rhino.RhinoDoc.ActiveDoc.Views.Redraw()
        
    except Exception as e:
        ghenv.Component.Message = "JSON Error: " + str(e)
else:
    if not run_btn: ghenv.Component.Message = "STATE: OFF"
    else: ghenv.Component.Message = "WAITING FOR DATA"

File: Components/Dash.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using Grasshopper.Kernel;
using Newtonsoft.Json;
using Rhino.Display;
using Rhino.Geometry;
using Enzyme;

namespace Enzyme.Components
{
    public class DashComponent : GH_Component
    {
        private bool _run = false;
        private string _jsonPayload = "";
        private bool _groupByBldg = false;
        private string _title = "MASTERPLAN SUMMARY";
        private string _suffix = "";
        private double _targetGlobal = 0.0;
        private string _targetJson = "";
        private double _size = 12.0;
        private int _anchor = 0;
        private string _fontFace = "Arial";
        private double _transparency = 0.8;
        private double _ox = 20.0;
        private double _oy = 20.0;
        private double _fitPadding = 10.0;
        
        private bool _subscribed = false;

        private List<DashDisplayItem> _displayData = new List<DashDisplayItem>();

        public DashComponent()
          : base("Masterplan Dashboard", "DASH",
              "Reads a summarized JSON payload from the OOP Masterplan Engine and renders a responsive HUD.",
              Enzyme.Utils.TabInfo.TabName, "Masterplan (Beta)")
        {
        }

        public override GH_Exposure Exposure => GH_Exposure.primary;

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                Enzyme.Utils.AutoWireHelper.WireToggle(this, document, 0, false, 210, -160);
                Enzyme.Utils.AutoWireHelper.WireToggle(this, document, 2, false, 210, -120);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 5, 0.0, 2.0, 0.0, 330, -80);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 7, 0.0, 24, 12.0, 330, -40);
                Enzyme.Utils.AutoWireHelper.WireValueList(this, document, 8, new string[]{"TL", "TR", "BL", "BR"}, new string[]{"0", "1", "2", "3"}, 300, 0);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 10, 0.0, 1.0, 0.8, 330, 40);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 11, 0.0, 40, 20.0, 330, 80);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 12, 0.0, 40, 20.0, 330, 120);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 13, 0.0, 20, 10.0, 330, 160);
            }
        }

        protected override void RegisterInputParams(GH_Component.GH_InputParamManager pManager)
        {
            pManager.AddBooleanParameter("Run", "run", "Enable HUD", GH_ParamAccess.item, false);
            pManager.AddTextParameter("JSON Payload", "JSON_Payload", "JSON Payload from OOP", GH_ParamAccess.item);
            pManager.AddBooleanParameter("Group By Bldg", "GroupByBldg", "Group By Building", GH_ParamAccess.item, false);
            pManager.AddTextParameter("Title", "title", "HUD Title", GH_ParamAccess.item, "MASTERPLAN SUMMARY");
            pManager.AddTextParameter("Suffix", "suffix", "Unit suffix", GH_ParamAccess.item, "");
            pManager.AddNumberParameter("Target Global", "TargetGlobal", "Target Global Area", GH_ParamAccess.item, 0.0);
            pManager.AddTextParameter("Target JSON", "TargetJSON", "JSON containing Targets", GH_ParamAccess.item, "");
            pManager.AddNumberParameter("Size", "size", "Font size", GH_ParamAccess.item, 12.0);
            pManager.AddIntegerParameter("Anchor", "anchor", "Anchor (0=TL, 1=TR, 2=BL, 3=BR)", GH_ParamAccess.item, 0);
            pManager.AddTextParameter("Font", "font", "Font Face", GH_ParamAccess.item, "Arial");
            pManager.AddNumberParameter("Transparency", "transparency", "Background Opacity (0.0 - 1.0)", GH_ParamAccess.item, 0.8);
            pManager.AddNumberParameter("Offset X", "ox", "Offset X", GH_ParamAccess.item, 20.0);
            pManager.AddNumberParameter("Offset Y", "oy", "Offset Y", GH_ParamAccess.item, 20.0);
            pManager.AddNumberParameter("Padding", "fit_padding", "Padding", GH_ParamAccess.item, 10.0);

            for (int i = 0; i < pManager.ParamCount; i++)
            {
                pManager[i].Optional = true;
            }
        }

        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
        {
            // No outputs, this component only draws to the HUD
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            // Fetch inputs
            DA.GetData(0, ref _run);
            DA.GetData(1, ref _jsonPayload);
            DA.GetData(2, ref _groupByBldg);
            DA.GetData(3, ref _title);
            DA.GetData(4, ref _suffix);
            DA.GetData(5, ref _targetGlobal);
            DA.GetData(6, ref _targetJson);
            DA.GetData(7, ref _size);
            DA.GetData(8, ref _anchor);
            DA.GetData(9, ref _fontFace);
            DA.GetData(10, ref _transparency);
            DA.GetData(11, ref _ox);
            DA.GetData(12, ref _oy);
            DA.GetData(13, ref _fitPadding);

            if (_run && !string.IsNullOrEmpty(_jsonPayload))
            {
                try
                {
                    _displayData.Clear();
                    
                    var data = JsonConvert.DeserializeObject<Dictionary<string, object>>(_jsonPayload);
                    
                    var progAreas = new Dictionary<string, double>();
                    if (data.ContainsKey("programs") && data["programs"] != null)
                        progAreas = JsonConvert.DeserializeObject<Dictionary<string, double>>(data["programs"].ToString());
                    
                    double totalArea = 0.0;
                    if (data.ContainsKey("total_area") && data["total_area"] != null)
                        totalArea = Convert.ToDouble(data["total_area"]);
                    
                    var bldgData = new Dictionary<string, Dictionary<string, object>>();
                    if (data.ContainsKey("buildings") && data["buildings"] != null)
                        bldgData = JsonConvert.DeserializeObject<Dictionary<string, Dictionary<string, object>>>(data["buildings"].ToString());

                    var cleanTargetData = new Dictionary<string, double>();
                    if (!string.IsNullOrEmpty(_targetJson))
                    {
                        try
                        {
                            var rawTargets = JsonConvert.DeserializeObject<Dictionary<string, object>>(_targetJson);
                            if (rawTargets != null)
                            {
                                foreach (var kvp in rawTargets)
                                {
                                    if (kvp.Value != null)
                                        cleanTargetData[kvp.Key.Trim().ToUpper()] = Convert.ToDouble(kvp.Value);
                                }
                            }
                        }
                        catch { }
                    }

                    if (!string.IsNullOrEmpty(_title))
                    {
                        _displayData.Add(new DashDisplayItem($"=== {_title.ToUpper()} ===", "", false));
                    }

                    if (_groupByBldg)
                    {
                        foreach (var kvp in bldgData)
                        {
                            string bName = kvp.Key;
                            var bStats = kvp.Value;
                            _displayData.Add(new DashDisplayItem($"--- {bName.ToUpper()} ---", "", false));
                            
                            if (bStats.ContainsKey("programs") && bStats["programs"] != null)
                            {
                                var pAreas = JsonConvert.DeserializeObject<Dictionary<string, double>>(bStats["programs"].ToString());
                                foreach (var pKvp in pAreas)
                                {
                                    string pName = pKvp.Key;
                                    double valArea = pKvp.Value;
                                    string lineLabel = $"  {pName}: ";
                                    string lineVal = $"{FormatNum(valArea)}{_suffix}";
                                    _displayData.Add(new DashDisplayItem(lineLabel, lineVal, false));
                                }
                            }
                            
                            double bTotArea = 0.0;
                            if (bStats.ContainsKey("total_area") && bStats["total_area"] != null)
                                bTotArea = Convert.ToDouble(bStats["total_area"]);
                                
                            string bTotVal = $"{FormatNum(bTotArea)}{_suffix}";
                            _displayData.Add(new DashDisplayItem("  SUBTOTAL: ", bTotVal, false));
                            _displayData.Add(new DashDisplayItem(" ", " ", false));
                        }

                        if (cleanTargetData.Count > 0)
                        {
                            _displayData.Add(new DashDisplayItem("--- TARGET TRACKING ---", "", false));
                            foreach (var pKvp in progAreas)
                            {
                                string progName = pKvp.Key;
                                double valArea = pKvp.Value;
                                string progNameUpper = progName.Trim().ToUpper();
                                double target = cleanTargetData.ContainsKey(progNameUpper) ? cleanTargetData[progNameUpper] : 0.0;
                                if (target > 0)
                                {
                                    double pct = (valArea / target) * 100.0;
                                    string lineLabel = $"  {progName}: ";
                                    string lineVal = $"{FormatNum(valArea)}{_suffix} / {FormatNum(target)}{_suffix} ({pct:F1}%)";
                                    _displayData.Add(new DashDisplayItem(lineLabel, lineVal, valArea > target));
                                }
                            }
                            _displayData.Add(new DashDisplayItem(" ", " ", false));
                        }

                        _displayData.Add(new DashDisplayItem(new string('=', 15), "", false));
                    }
                    else
                    {
                        foreach (var pKvp in progAreas)
                        {
                            string progName = pKvp.Key;
                            double valArea = pKvp.Value;
                            string progNameUpper = progName.Trim().ToUpper();
                            double target = cleanTargetData.ContainsKey(progNameUpper) ? cleanTargetData[progNameUpper] : 0.0;
                            
                            string lineLabel = $"{progName.ToUpper()}: ";
                            string lineVal = $"{FormatNum(valArea)}{_suffix}";
                            bool overTarget = false;
                            
                            if (target > 0)
                            {
                                double pct = (valArea / target) * 100.0;
                                lineVal += $" / {FormatNum(target)}{_suffix} ({pct:F1}%)";
                                if (valArea > target) overTarget = true;
                            }
                            
                            _displayData.Add(new DashDisplayItem(lineLabel, lineVal, overTarget));
                        }
                        _displayData.Add(new DashDisplayItem(new string('-', 15), "", false));
                    }

                    bool totIsRed = false;
                    string totStr = $"{FormatNum(totalArea)}{_suffix}";
                    
                    if (_targetGlobal > 0)
                    {
                        double totPct = (totalArea / _targetGlobal) * 100.0;
                        totStr += $" [Target: {FormatNum(_targetGlobal)}{_suffix} ({totPct:F1}%)]";
                        if (totalArea > _targetGlobal) totIsRed = true;
                    }
                    
                    _displayData.Add(new DashDisplayItem("TOTAL AREA: ", totStr, totIsRed));

                    if (!_subscribed)
                    {
                        DisplayPipeline.DrawForeground += OnDrawForeground;
                        _subscribed = true;
                    }

                    string modeMsg = _groupByBldg ? "Per-Building Mode" : "Global Mode";
                    this.Message = $"HUD ACTIVE\n{modeMsg}";
                }
                catch (Exception ex)
                {
                    this.Message = "JSON Error: " + ex.Message;
                }
            }
            else
            {
                if (!_run)
                    this.Message = "STATE: OFF";
                else
                    this.Message = "WAITING FOR DATA";
                
                Unsubscribe();
            }
        }

        private void OnDrawForeground(object sender, DrawEventArgs e)
        {
            if (!_run || _displayData.Count == 0) return;

            int vpW = e.Viewport.Bounds.Width;
            int vpH = e.Viewport.Bounds.Height;
            double lineH = _size * 1.6;
            double titleGap = 20.0;

            int maxLabelChars = 10;
            foreach (var d in _displayData)
                if (d.Label.Length > maxLabelChars) maxLabelChars = d.Label.Length;

            double valueStartOffset = maxLabelChars * (_size * 0.6);

            int maxTotalChars = 10;
            foreach (var d in _displayData)
            {
                int len = d.Label.Length + d.Val.Length;
                if (len > maxTotalChars) maxTotalChars = len;
            }
            
            double boxW = (maxTotalChars * _size * 0.55) + (_fitPadding * 3.0);

            int titleCount = 0;
            foreach (var d in _displayData)
            {
                if (d.Label.Contains("---") || d.Label.Contains("==="))
                    titleCount++;
            }

            double boxH = (_displayData.Count * lineH) + (titleCount * titleGap) + (_fitPadding * 2.5);

            double x = 0, y = 0;
            if (_anchor == 1) { x = vpW - boxW - _ox; y = _oy; }
            else if (_anchor == 2) { x = _ox; y = vpH - boxH - _oy; }
            else if (_anchor == 3) { x = vpW - boxW - _ox; y = vpH - boxH - _oy; }
            else { x = _ox; y = _oy + 35; }

            double transparencyClamped = Math.Max(0.0, Math.Min(1.0, _transparency));
            int alpha = (int)(transparencyClamped * 255);
            Color bg = Color.FromArgb(alpha, 25, 25, 25);
            Color white = Color.White;
            Color warningRed = Color.OrangeRed;

            var rect = new Rectangle((int)x, (int)y, (int)boxW, (int)boxH);
            e.Display.Draw2dRectangle(rect, bg, 1, bg);

            double currentY = y + _fitPadding;
            foreach (var item in _displayData)
            {
                if (string.IsNullOrWhiteSpace(item.Label) && string.IsNullOrWhiteSpace(item.Val))
                {
                    currentY += lineH;
                    continue;
                }

                e.Display.Draw2dText(item.Label, white, new Point2d(x + _fitPadding, currentY), false, (int)_size, _fontFace);
                if (!item.Label.Contains("---") && !item.Label.Contains("==="))
                {
                    var valP = new Point2d(x + _fitPadding + valueStartOffset, currentY);
                    Color color = item.IsRed ? warningRed : white;
                    e.Display.Draw2dText(item.Val, color, valP, false, (int)_size, _fontFace);
                }

                currentY += (item.Label.Contains("---") || item.Label.Contains("===")) ? (lineH + titleGap) : lineH;
            }
        }

        private string FormatNum(double n)
        {
            return n.ToString("N2");
        }

        private void Unsubscribe()
        {
            if (_subscribed)
            {
                DisplayPipeline.DrawForeground -= OnDrawForeground;
                _subscribed = false;
            }
        }

        public override void RemovedFromDocument(GH_Document document)
        {
            Unsubscribe();
            base.RemovedFromDocument(document);
        }
        
        public override void DocumentContextChanged(GH_Document document, GH_DocumentContext context)
        {
            if (context == GH_DocumentContext.Close || context == GH_DocumentContext.Unloaded)
            {
                Unsubscribe();
            }
            base.DocumentContextChanged(document, context);
        }

        protected override Bitmap Icon => IconLoader.Load("DASH.png");

        public override Guid ComponentGuid => new Guid("B453E8A0-E3A9-4A9E-B19F-D7C6B372F26C");

        private class DashDisplayItem
        {
            public string Label { get; }
            public string Val { get; }
            public bool IsRed { get; }

            public DashDisplayItem(string label, string val, bool isRed)
            {
                Label = label ?? "";
                Val = val ?? "";
                IsRed = isRed;
            }
        }
    }
}
